storage/tianmu/core/mysql_expression.cpp

```diff
@@ -618,22 +621,14 @@ bool operator==(Item const &l_, Item const &r_) {
           same = same && l->arg_count == r->arg_count;
           for (uint i = 0; same && (i < l->arg_count); ++i) same = same && (*l->arguments()[i] == *r->arguments()[i]);
 
-          // Item_func* lll = (Item_func*)&l;
-          // Item_func* mmm = (Item_func*)&r;
-
-          // bool x = l->const_item();
-          // bool y = r->const_item();
-          // longlong zzz = lll->val_int_result();
-          // longlong vvv = mmm->val_int_result();
           same = same && (l->const_item() == r->const_item());
           if (same && l->const_item())
             same = ((Item_func *)&l_)->val_int() == ((Item_func *)&r_)->val_int();
-          if (dynamic_cast<const Item_date_add_interval *>(&l_)) {
-            const Item_date_add_interval *l = static_cast<const Item_date_add_interval *>(&l_);
-            const Item_date_add_interval *r = static_cast<const Item_date_add_interval *>(&r_);
-            same = same && dynamic_cast<const Item_date_add_interval *>(&r_);
-            same = same && ((l->int_type == r->int_type) && (l->date_sub_interval == r->date_sub_interval));
-          }
+
+          const Item_date_add_interval *l = static_cast<const Item_date_add_interval *>(&l_);
```

Comment:
uses `down_cast` not `static_cast` directly.

Reply:
fixed it

Reply:
there can not use down_cast,  will lead to some mtr cases crash  in debug mode,  as below case:
```
CREATE TABLE `select_base_with_primary_key` (
  `pk` bigint(11) NOT NULL AUTO_INCREMENT,
  `varchar_test` varchar(255) DEFAULT NULL,
  `integer_test` int(11) DEFAULT NULL,
  `char_test` char(255) DEFAULT NULL,
  `tinyint_test` tinyint(4) DEFAULT NULL,
  `tinyint_1bit_test` tinyint(1) DEFAULT NULL,
  `smallint_test` smallint(6) DEFAULT NULL,
  `mediumint_test` mediumint(9) DEFAULT NULL,
  `bigint_test` bigint(20) DEFAULT NULL,
  `double_test` double DEFAULT NULL,
  `decimal_test` decimal(10,0) DEFAULT NULL,
  `date_test` date DEFAULT NULL,
  `time_test` time DEFAULT NULL,
  `datetime_test` datetime DEFAULT NULL,
  `timestamp_test` timestamp NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
  `year_test` year(4) DEFAULT NULL,
  PRIMARY KEY (`pk`)
) ENGINE=TIANMU DEFAULT CHARSET=gbk;
insert into select_base_with_primary_key values (11, 'abcde', 123, 'char', 3, 1, 23, 32, 332, 1.4, 25, '2017-11-11', '11:11:11', '2017-11-11 11:11:11','2017-11-11 11:11:11', 2017);
select 'aa' as ag_column1,  !( tinyint_1bit_test )as ag_column2, 1 as ag_column3,min((smallint_test +  smallint_test) =11)as ag_column_fun  from select_base_with_primary_key    GROUP BY  ag_column1,ag_column2  asc,ag_column3   having ag_column_fun<= 37 order by  ag_column1    ,ag_column2    asc,ag_column3    asc,ag_column_fun;
```
stack：
```
down_cast<Item_date_add_interval const*, Item const*>(const Item * arg) (\opt\github\stonedb57\include\template_utils.h:85)
Tianmu::core::operator==(const Item & l_, const Item & r_) (\opt\github\stonedb57\storage\tianmu\core\mysql_expression.cpp:628)
Tianmu::core::MysqlExpression::operator==(const Tianmu::core::MysqlExpression * const this, const Tianmu::core::MysqlExpression & other) (\opt\github\stonedb57\storage\tianmu\core\mysql_expression.cpp:684)
Tianmu::core::Query::VirtualColumnAlreadyExists(Tianmu::core::Query * const this, const Tianmu::core::TabID & tmp_table, Tianmu::core::MysqlExpression * expression) (\opt\github\stonedb57\storage\tianmu\core\query.cpp:139)
Tianmu::core::Query::AddColumnForMysqlExpression(Tianmu::core::Query * const this, Tianmu::core::MysqlExpression * mysql_expression, const Tianmu::core::TabID & tmp_table, const char * alias, const Tianmu::common::ColOperation oper, const bool distinct, bool group_by) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:849)
Tianmu::core::Query::AddFields(Tianmu::core::Query * const this, List<Item> & fields, const Tianmu::core::TabID & tmp_table, const bool group_by_clause, int & num_of_added_fields, bool ignore_minmax, bool & aggregation_used) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:474)
Tianmu::core::Query::Compile(Tianmu::core::Query * const this, Tianmu::core::CompiledQuery * compiled_query, SELECT_LEX * selects_list, SELECT_LEX * last_distinct, Tianmu::core::TabID * res_tab, bool ignore_limit, Item * left_expr_for_subselect, Tianmu::common::Operator * oper_for_subselect, bool ignore_minmax, bool for_subq_in_where) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:1112)
Tianmu::core::Engine::Execute(Tianmu::core::Engine * const this, THD * thd, LEX * lex, Query_result * result_output, SELECT_LEX_UNIT * unit_for_union) (\opt\github\stonedb57\storage\tianmu\core\engine_execute.cpp:436)
Tianmu::core::Engine::HandleSelect(Tianmu::core::Engine * const this, THD * thd, LEX * lex, Query_result *& result, ulong setup_tables_done_option, int & res, int & optimize_after_tianmu, int & tianmu_free_join, int with_insert) (\opt\github\stonedb57\storage\tianmu\core\engine_execute.cpp:238)
Tianmu::handler::ha_my_tianmu_query(THD * thd, LEX * lex, Query_result *& result_output, ulong setup_tables_done_option, int & res, int & optimize_after_tianmu, int & tianmu_free_join, int with_insert) (\opt\github\stonedb57\storage\tianmu\handler\ha_my_tianmu.cpp:88)
execute_sqlcom_select(THD * thd, TABLE_LIST * all_tables) (\opt\github\stonedb57\sql\sql_parse.cc:5184)
mysql_execute_command(THD * thd, bool first_level) (\opt\github\stonedb57\sql\sql_parse.cc:2831)
mysql_parse(THD * thd, Parser_state * parser_state) (\opt\github\stonedb57\sql\sql_parse.cc:5622)
dispatch_command(THD * thd, const COM_DATA * com_data, enum_server_command command) (\opt\github\stonedb57\sql\sql_parse.cc:1495)
do_command(THD * thd) (\opt\github\stonedb57\sql\sql_parse.cc:1034)
handle_connection(void * arg) (\opt\github\stonedb57\sql\conn_handler\connection_handler_per_thread.cc:313)
```
breakpiont
```
template<typename Target, typename Source>
inline Target down_cast(Source arg)
{
  assert(NULL != dynamic_cast<Target>(arg));
  return static_cast<Target>(arg);
}

```

Reply:
`l` and `r ` type is `Item_date_add_interval` :
```
Item_date_add_interval -> Item_temporal_hybrid_func -> Item_temporal_hybrid_func
 -> Item_str_func -> Item_str_func -> Item_func -> Item_result_field
 -> Itemclass Item_func -> Item_result_field
```

Reply:
so ，cann't  use down_case